### ISSUE TITLE
Eventchannel shows coded fields instead of text

### ISSUE THREAD

**Issue by cristgl** [wazuh/wazuh#3242]
|Wazuh version|Install type|Install method|Platform|
|---|---|---|---|
| 3.9.0-3917 | Manager/Agent | Packages/Sources | Windows |

```
** Alert 1551788915.3574067: - windows,
2019 Mar 05 13:28:35 (win) any->EventChannel
Rule: 70002 (level 9) -> 'Windows Defender scan detected the malware HackTool:Win32/Mimikatz.E' 
{"win":{"system":{"providerName":"Microsoft-Windows-Windows Defender","providerGuid":"{11CD958A-C507-4EF3-B3F2-5FD9DFBD2C78}","eventID":"1116","version":"0","level":"3","task":"0","opcode":"0","keywords":"0x8000000000000000","systemTime":"2019-03-05T12:28:34.314600800Z","eventRecordID":"272","processID":"1524","threadID":"1972","channel":"Microsoft-Windows-Windows Defender/Operational","computer":"pcname","severityValue":"WARNING","message":"Windows Defender has detected malware or other potentially unwanted software."},"eventdata":{"product Name":"%%827","product Version":"4.10.209.0","detection ID":"{4DF8EA0A-8B15-48BC-8407-289C7C52F330}","detection Time":"2019-03-05T11:59:06.304Z","unused":"","unused2":"","threat ID":"2147730094","threat Name":"HackTool:Win32/Mimikatz.E","severity ID":"4","severity Name":"High","category ID":"34","category Name":"Tool","fWLink":"http://go.microsoft.com/fwlink/?linkid=37020&amp;amp;amp;amp;amp;amp;amp;amp;name=HackTool:Win32/Mimikatz.E&amp;amp;amp;amp;amp;amp;amp;amp;threatid=2147730094&amp;amp;amp;amp;amp;amp;amp;amp;enterprise=0","status Code":"1","status Description":"","state":"1","source ID":"1","source Name":"%%815","process Name":"C:\\Windows\\explorer.exe","detection User":"pcname\\user","unused3":"","path":"containerfile:_C:\\Users\\user\\AppData\\Local\\Mozilla\\Firefox\\Profiles\\6bw9jpsh.default\\cache2\\entries\\62E65F9830CF8AACC514E164B99FCBE9E3BB6609;file:_C:\\Users\\user\\AppData\\Local\\Mozilla\\Firefox\\Profiles\\6bw9jpsh.default\\cache2\\entries\\62E65F9830CF8AACC514E164B99FCBE9E3BB6609-&amp;amp;amp;amp;amp;amp;amp;gt;x64/mimikatz.exe;file:_C:\\Users\\user\\Desktop\\mimikatz_trunk\\x64\\mimikatz.exe;file:_C:\\Users\\user\\Downloads\\mimikatz_trunk\\x64\\mimikatz.exe;process:_pid:3600,ProcessStart:131962604780668201","origin ID":"1","origin Name":"%%845","execution ID":"3","execution Name":"%%848","type ID":"0","type Name":"%%822","pre Execution Status":"0","action ID":"9","action Name":"%%887","unused4":"","error Code":"0x00000000","error Description":"The operation completed successfully.","unused5":"","post Clean Status":"0","additional Actions ID":"0","additional Actions String":"No additional actions required","remediation User":"","unused6":"","signature Version":"AV: 1.289.481.0, AS: 1.289.481.0, NIS: 119.0.0.0","engine Version":"AM: 1.1.15700.9, NIS: 2.1.14600.4"}}} 
win.system.providerName: Microsoft-Windows-Windows Defender 
win.system.providerGuid: {11CD958A-C507-4EF3-B3F2-5FD9DFBD2C78} 
win.system.eventID: 1116 
win.system.version: 0 
win.system.level: 3 
win.system.task: 0 
win.system.opcode: 0 
win.system.keywords: 0x8000000000000000 
win.system.systemTime: 2019-03-05T12:28:34.314600800Z 
win.system.eventRecordID: 272 
win.system.processID: 1524
win.system.threadID: 1972
win.system.channel: Microsoft-Windows-Windows Defender/Operational 
win.system.computer: pcname 
win.system.severityValue: WARNING 
win.system.message: Windows Defender has detected malware or other potentially unwanted software. 
win.eventdata.product Name: %%827 
win.eventdata.product Version: 4.10.209.0 
win.eventdata.detection ID: {4DF8EA0A-8B15-48BC-8407-289C7C52F330} 
win.eventdata.detection Time: 2019-03-05T11:59:06.304Z 
win.eventdata.threat ID: 2147730094 
win.eventdata.threat Name: HackTool:Win32/Mimikatz.E 
win.eventdata.severity ID: 4 win.eventdata.severity Name: High 
win.eventdata.category ID: 34 win.eventdata.category Name: Tool
win.eventdata.fWLink: http://go.microsoft.com/fwlink/?linkid=37020name=HackTool:Win32/Mimikatz.Ethreatid=2147730094enterprise=0 
win.eventdata.status Code: 1 
win.eventdata.state: 1 
win.eventdata.source ID: 1 
win.eventdata.source Name: %%815 
win.eventdata.process Name: C:\Windows\explorer.exe 
win.eventdata.detection User: pcname\user 
win.eventdata.path: containerfile:_C:\Users\user\AppData\Local\Mozilla\Firefox\Profiles\6bw9jpsh.default\cache2\entries\62E65F9830CF8AACC514E164B99FCBE9E3BB6609;file:_C:\Users\user\AppData\Local\Mozilla\Firefox\Profiles\6bw9jpsh.default\cache2\entries\62E65F9830CF8AACC514E164B99FCBE9E3BB6609-&amp;amp;amp;amp;amp;amp;amp;gt;x64/mimikatz.exe;file:_C:\Users\user\Desktop\mimikatz_trunk\x64\mimikatz.exe;file:_C:\Users\user\Downloads\mimikatz_trunk\x64\mimikatz.exe;process:_pid:3600,ProcessStart:131962604780668201 
win.eventdata.origin ID: 1 
win.eventdata.origin Name: %%845 
win.eventdata.execution ID: 3 
win.eventdata.execution Name: %%848 
win.eventdata.type ID: 0 
win.eventdata.type Name: %%822 
win.eventdata.pre Execution Status: 0 
win.eventdata.action ID: 9 
win.eventdata.action Name: %%887 
win.eventdata.error Code: 0x00000000 
win.eventdata.error Description: The operation completed successfully. 
win.eventdata.post Clean Status: 0 
win.eventdata.additional Actions ID: 0 
win.eventdata.additional Actions String: No additional actions required 
win.eventdata.signature Version: AV: 1.289.481.0, AS: 1.289.481.0, NIS: 119.0.0.0 win.eventdata.engine Version: AM: 1.1.15700.9, NIS: 2.1.14600.4
```

The alert from above has some fields which are coded. They should be converted to text:

win.eventdata.product Name: %%827
win.eventdata.source Name: %%815
win.eventdata.origin Name: %%845
win.eventdata.execution Name: %%848
win.eventdata.type Name: %%822
win.eventdata.action Name: %%887

**Comment by cristgl** [comment]
There is no Windows API function that return a human-readable value from a code.

The library file `msobj.dll` contains, for each Windows version, those codes and their values, but not for every channel, only for the Security one. This does not help with the solution for this issue, as the channel in this case is the one from Windows Defender. Nevertheless, there are plenty of codes in that file, that means that much more events other than the referenced in this [issue](https://github.com/wazuh/wazuh/issues/2937) are returning code fields.

I'm trying to notice if those events are common, if they are, it could be useful to extract the data from the _msobj.dll_ file. Probably, extracting the information in that file could be done with the Windows API functions `LoadLibrary` and `GetProcAddress`, but I'm not sure, as for the last one, the two parameters are the library and a variable or function name, could a code field be passed?

Following with the issue, the code values aren't either in the `Details` section of Event Viewer, as it was in the issue referenced above. It is getting complicated to know which values these fields have.

**Comment by cristgl** [comment]
I have been trying to get the information from the _msobj.dll_ file, just to get the entire dictionary of the Security channel codes from each Windows system.
I have used the function `LoadLibrary` to load the library and `GetProcAddress` to get each code's human-readable value. For example, the event 4688 has the field `TokenElevationType` with value `%%1937`. 

Now, the question is, how can I know which type of value will be returned when I call the `GetProcAddress` function passing it 1937 as parameter? I have tried to check it with `char *`, but `null` is returned. I can't find any information about the variable types included in this file.

**Comment by cristgl** [comment]
I have finally manage to get the codes with their _human-readable_ values, there are 400000 only for Windows 8's Security channel. These codes are retrieved at the agent's side, which involves parsing the XML message from each event to translate each code included in them. There are plenty of events being recorded in the Windows' eventlog each minute, this would totally affect performance from the agent. Also, it cannot be done at the manager's side as it is necessary to read the _msobjs.dll_ file, which can only be accessed on Windows.

**Comment by cristgl** [comment]
To find out which is the value for the specific code that may come from the event there are two alternatives:

- Searching for the codes that may be included and loading the library to find them for each event.
In this case we should find each code, save its key (for example, for the field `<Data Name="LayerName">%%14608</Data>`, the key would be `LayerName`), rename it to `keyValue` and add the translation of its code, which can be one or more.

- Returning the entire message field. Eventchannel was not returning this information because it usually repeats some fields, for example, the next event.
This is the event in XML:
```
- <Event xmlns="http://schemas.microsoft.com/win/2004/08/events/event">
- <System>
  <Provider Name="Microsoft-Windows-Security-Auditing" Guid="{54849625-5478-4994-A5BA-3E3B0328C30D}" /> 
  <EventID>5158</EventID> 
  <Version>0</Version> 
  <Level>0</Level> 
  <Task>12810</Task> 
  <Opcode>0</Opcode> 
  <Keywords>0x8020000000000000</Keywords> 
  <TimeCreated SystemTime="2019-05-10T14:17:53.395364800Z" /> 
  <EventRecordID>39272</EventRecordID> 
  <Correlation /> 
  <Execution ProcessID="4" ThreadID="892" /> 
  <Channel>Security</Channel> 
  <Computer>WIN-9MQ21JGECIU</Computer> 
  <Security /> 
  </System>
- <EventData>
  <Data Name="ProcessId">1108</Data> 
  <Data Name="Application">\device\harddiskvolume2\program files (x86)\ossec-agent\ossec-agent.exe</Data> 
  <Data Name="SourceAddress">0.0.0.0</Data> 
  <Data Name="SourcePort">49549</Data> 
  <Data Name="Protocol">17</Data> 
  <Data Name="FilterRTID">0</Data> 
  <Data Name="LayerName">%%14608</Data> 
  <Data Name="LayerRTID">36</Data> 
  </EventData>
  </Event>
```

This is the message field, which can be seen at the "General" tab.
![imagen](https://user-images.githubusercontent.com/14914658/57535894-5bb36f80-7343-11e9-8292-1b023f3ee1d4.png)

It can be seen that every field contained in the `EventData` block is also included in the message. If we get the message also, the final event would contain some duplicated data and the message written in the default language from the Windows system.

**Comment by cristgl** [comment]
Following with the second alternative proposed in the last comment, I have obtained some information.
The message is already being retrieved in the right format, as it was seen in the last image, it has the spaces and tabs that separate each field. 
At the time this message is added to the JSON object, it loses its format and adds some '\n', '\t' and '\r' bytes to represent these spaces. If we are using JSON format, there is no possibility to change these bytes for spaces, which means the entire message would be written in just one line with some spaces, as the whole event is then parsed to a JSON object.

**Comment by cristgl** [comment]
At the end, I managed to get the message with the right format. At the agent side, I added the message at the end of the JSON event that is sent to the manager, but not as a part of it. At the manager side, this message is obtained and filtered and it can be added to the final JSON object. Below is a sample alert with the formatted message; also mention that the 'Message:' part will be deleted, as well as some other changes from the code will be modified.

![image (3)](https://user-images.githubusercontent.com/14914658/57684748-b1836280-7636-11e9-9c14-ff693f6d8b64.png)

**Comment by cristgl** [comment]
Working at branch `fix-3242`.

This is a large event sample of how it has ended to be:
```
** Alert 1557831160.918497: - windows, windows_security,authentication_success,pci_dss_10.2.5,gpg13_7.1,gpg13_7.2,gdpr_IV_32.2,
2019 May 14 12:52:40 (win) any->EventChannel
Rule: 60118 (level 3) -> 'Windows Workstation Logon Success'
{"win":{"system":{"providerName":"Microsoft-Windows-Security-Auditing","providerGuid":"{54849625-5478-4994-A5BA-3E3B0328C30D}","eventID":"4624","version":"1","level":"0","task":"12544","opcode":"0","keywords":"0x8020000000000000","systemTime":"2019-05-14T10:52:39.017777700Z","eventRecordID":"15743080","processID":"568","threadID":"940","channel":"Security","computer":"pcname","severityValue":"AUDIT_SUCCESS","message":"An account was successfully logged on.\r\n\r\nSubject:\r\n\tSecurity ID:\t\tS-1-5-18\r\n\tAccount Name:\t\tPCNAME$\r\n\tAccount Domain:\t\tWORKGROUP\r\n\tLogon ID:\t\t0x3E7\r\n\r\nLogon Type:\t\t\t2\r\n\r\nImpersonation Level:\t\tImpersonation\r\n\r\nNew Logon:\r\n\tSecurity ID:\t\tS-1-5-21-3416783167-2895274904-3428114391-1001\r\n\tAccount Name:\t\tuser\r\n\tAccount Domain:\t\tpcname\r\n\tLogon ID:\t\t0x5BF070\r\n\tLogon GUID:\t\t{00000000-0000-0000-0000-000000000000}\r\n\r\nProcess Information:\r\n\tProcess ID:\t\t0x1380\r\n\tProcess Name:\t\tC:\\Windows\\System32\\winlogon.exe\r\n\r\nNetwork Information:\r\n\tWorkstation Name:\tPCNAME\r\n\tSource Network Address:\t127.0.0.1\r\n\tSource Port:\t\t0\r\n\r\nDetailed Authentication Information:\r\n\tLogon Process:\t\tUser32 \r\n\tAuthentication Package:\tNegotiate\r\n\tTransited Services:\t-\r\n\tPackage Name (NTLM only):\t-\r\n\tKey Length:\t\t0\r\n\r\nThis event is generated when a logon session is created. It is generated on the computer that was accessed.\r\n\r\nThe subject fields indicate the account on the local system which requested the logon. This is most commonly a service such as the Server service, or a local process such as Winlogon.exe or Services.exe.\r\n\r\nThe logon type field indicates the kind of logon that occurred. The most common types are 2 (interactive) and 3 (network).\r\n\r\nThe New Logon fields indicate the account for whom the new logon was created, i.e. the account that was logged on.\r\n\r\nThe network fields indicate where a remote logon request originated. Workstation name is not always available and may be left blank in some cases.\r\n\r\nThe impersonation level field indicates the extent to which a process in the logon session can impersonate.\r\n\r\nThe authentication information fields provide detailed information about this specific logon request.\r\n\t- Logon GUID is a unique identifier that can be used to correlate this event with a KDC event.\r\n\t- Transited services indicate which intermediate services have participated in this logon request.\r\n\t- Package name indicates which sub-protocol was used among the NTLM protocols.\r\n\t- Key length indicates the length of the generated session key. This will be 0 if no session key was requested."},"eventdata":{"subjectUserSid":"S-1-5-18","subjectUserName":"PCNAME$","subjectDomainName":"WORKGROUP","subjectLogonId":"0x3e7","targetUserSid":"S-1-5-21-3416783167-2895274904-3428114391-1001","targetUserName":"user","targetDomainName":"pcname","targetLogonId":"0x5bf070","logonType":"2","logonProcessName":"User32","authenticationPackageName":"Negotiate","workstationName":"PCNAME","logonGuid":"{00000000-0000-0000-0000-000000000000}","keyLength":"0","processId":"0x1380","processName":"C:\\Windows\\System32\\winlogon.exe","ipAddress":"127.0.0.1","ipPort":"0","impersonationLevel":"%%1833"}}}
win.system.providerName: Microsoft-Windows-Security-Auditing
win.system.providerGuid: {54849625-5478-4994-A5BA-3E3B0328C30D}
win.system.eventID: 4624
win.system.version: 1
win.system.level: 0
win.system.task: 12544
win.system.opcode: 0
win.system.keywords: 0x8020000000000000
win.system.systemTime: 2019-05-14T10:52:39.017777700Z
win.system.eventRecordID: 15743080
win.system.processID: 568
win.system.threadID: 940
win.system.channel: Security
win.system.computer: pcname
win.system.severityValue: AUDIT_SUCCESS
win.system.message: An account was successfully logged on.

Subject:
	Security ID:		S-1-5-18
	Account Name:		PCNAME$
	Account Domain:		WORKGROUP
	Logon ID:		0x3E7

Logon Type:			2

Impersonation Level:		Impersonation

New Logon:
	Security ID:		S-1-5-21-3416783167-2895274904-3428114391-1001
	Account Name:		user
	Account Domain:		pcname
	Logon ID:		0x5BF070
	Logon GUID:		{00000000-0000-0000-0000-000000000000}

Process Information:
	Process ID:		0x1380
	Process Name:		C:\Windows\System32\winlogon.exe

Network Information:
	Workstation Name:	PCNAME
	Source Network Address:	127.0.0.1
	Source Port:		0

Detailed Authentication Information:
	Logon Process:		User32 
	Authentication Package:	Negotiate
	Transited Services:	-
	Package Name (NTLM only):	-
	Key Length:		0

This event is generated when a logon session is created. It is generated on the computer that was accessed.

The subject fields indicate the account on the local system which requested the logon. This is most commonly a service such as the Server service, or a local process such as Winlogon.exe or Services.exe.

The logon type field indicates the kind of logon that occurred. The most common types are 2 (interactive) and 3 (network).

The New Logon fields indicate the account for whom the new logon was created, i.e. the account that was logged on.

The network fields indicate where a remote logon request originated. Workstation name is not always available and may be left blank in some cases.

The impersonation level field indicates the extent to which a process in the logon session can impersonate.

The authentication information fields provide detailed information about this specific logon request.
	- Logon GUID is a unique identifier that can be used to correlate this event with a KDC event.
	- Transited services indicate which intermediate services have participated in this logon request.
	- Package name indicates which sub-protocol was used among the NTLM protocols.
	- Key length indicates the length of the generated session key. This will be 0 if no session key was requested.
win.eventdata.subjectUserSid: S-1-5-18
win.eventdata.subjectUserName: PCNAME$
win.eventdata.subjectDomainName: WORKGROUP
win.eventdata.subjectLogonId: 0x3e7
win.eventdata.targetUserSid: S-1-5-21-3416783167-2895274904-3428114391-1001
win.eventdata.targetUserName: user
win.eventdata.targetDomainName: pcname
win.eventdata.targetLogonId: 0x5bf070
win.eventdata.logonType: 2
win.eventdata.logonProcessName: User32
win.eventdata.authenticationPackageName: Negotiate
win.eventdata.workstationName: PCNAME
win.eventdata.logonGuid: {00000000-0000-0000-0000-000000000000}
win.eventdata.keyLength: 0
win.eventdata.processId: 0x1380
win.eventdata.processName: C:\Windows\System32\winlogon.exe
win.eventdata.ipAddress: 127.0.0.1
win.eventdata.ipPort: 0
win.eventdata.impersonationLevel: %%1833
```

I'm doing some tests and I have encountered a segmentation fault whenever an event comes with an empty message. It's resulting difficult to reproduce as it was launched when a random event from the system was generated.

**Comment by cristgl** [comment]
The segmentation fault commented above has been fixed in this [commit](https://github.com/wazuh/wazuh/commit/eb3c26c71f85c21426b1cca0e1c9a9e28b37a1d0) This is the use case:
The channel Sysmon was being monitored, it was reporting every generated event, I uninstalled Sysmon and the message from this channel could not be obtained, as there wasn't any channel called Sysmon. The code at the agent side was filling the final event sent to the manager when the message is retrieved, that is why the agent sent a null as event and the manager stopped with a segmentation fault.

**Comment by cristgl** [comment]
With the next configuration:
```
<localfile>
    <location>Application</location>
    <log_format>eventchannel</log_format>
  </localfile>

  <localfile>
    <location>Security</location>
    <log_format>eventchannel</log_format>
  </localfile>

  <localfile>
    <location>System</location>
    <log_format>eventchannel</log_format>
  </localfile>
  
  <localfile>
      <location>Windows PowerShell</location>
      <log_format>eventchannel</log_format>
  </localfile>
    
  <localfile>
      <location>Microsoft-Windows-Application-Experience/Program-Inventory</location>
      <log_format>eventchannel</log_format>
  </localfile>
  
  <localfile>
      <location>Microsoft-Windows-DNS-Client/Operational</location>
      <log_format>eventchannel</log_format>
  </localfile>
  
  <localfile>
      <location>Microsoft-Windows-DriverFrameworks-UserMode/Operational</location>
      <log_format>eventchannel</log_format>
  </localfile>
  
  <localfile>
      <location>Microsoft-Windows-Sysmon/Operational</location>
      <log_format>eventchannel</log_format>
  </localfile>
    
  <localfile>
      <location>Microsoft-Windows-TaskScheduler/Operational</location>
      <log_format>eventchannel</log_format>
  </localfile>
    
  <localfile>
      <location>Microsoft-Windows-TerminalServices-RemoteConnectionManager/Operational</location>
      <log_format>eventchannel</log_format>
  </localfile>
  
  <localfile>
      <location>Microsoft-Windows-Windows Firewall With Advanced Security/Firewall</location>
      <log_format>eventchannel</log_format>
  </localfile>
```

The buffer is full. When this happens, the memory used raises from 13~14 MB to 22~23 MB. When it gets back to a size under 70%, it recovers to 13~14 MB.

Agent side:
```
2019/06/03 07:02:11 ossec-agent: WARNING: Agent buffer at 90 %.
2019/06/03 07:02:11 ossec-agent: WARNING: Agent buffer is full: Events may be lost.
2019/06/03 07:02:26 ossec-agent: WARNING: Agent buffer is flooded: Producing too many events.
2019/06/03 07:03:05 ossec-agent: INFO: Agent buffer is under 70 %. Working properly again.
2019/06/03 07:03:14 ossec-agent: WARNING: Agent buffer at 90 %.
2019/06/03 07:03:39 ossec-agent: INFO: Agent buffer is under 70 %. Working properly again.
2019/06/03 07:10:24 ossec-agent: WARNING: Agent buffer at 90 %.
2019/06/03 07:10:25 ossec-agent: WARNING: Agent buffer is full: Events may be lost.
2019/06/03 07:10:40 ossec-agent: WARNING: Agent buffer is flooded: Producing too many events.
2019/06/03 07:14:17 ossec-agent: INFO: Agent buffer is under 70 %. Working properly again.
2019/06/03 07:14:34 ossec-agent: WARNING: Agent buffer at 90 %.
2019/06/03 07:14:40 ossec-agent: WARNING: Agent buffer is full: Events may be lost.
2019/06/03 07:14:55 ossec-agent: WARNING: Agent buffer is flooded: Producing too many events.
2019/06/03 07:16:25 ossec-agent: INFO: Agent buffer is under 70 %. Working properly again.
```

Manager side:
`2019/06/03 16:14:44 ossec-analysisd: WARNING: Windows eventchannel decoder queue is full.`

**Comment by chemamartinez** [comment]
Fixed at #3320
